File: Sliding_Window/3_Longest_Substring_Without_Repeating_Characters/solution3.cpp
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> charSet;
        int l = 0, res = 0;

        for (int r = 0; r < s.length(); ++r) {
            while (charSet.find(s[r]) != charSet.end()) {
                charSet.erase(s[l]);
                l++;
            }
            charSet.insert(s[r]);
            res = max(res, r - l + 1);
        }
        
        return res;
    }
};

// Runtime1: 24ms, beats 27.86%
// Memory1: 14.38MB, beats 38.16%

// Runtime2: 20ms, beats 38.78%
// Memory2: 14.47MB, beats 36.94%

// Runtime3: 24ms, beats 27.86%
// Memory3: 14.57MB, beats 33.48%

File: Sliding_Window/3_Longest_Substring_Without_Repeating_Characters/solution3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charSet = set()
        l = 0
        res = 0

        for r in range(len(s)):
            while s[r] in charSet:
                charSet.remove(s[l])
                l += 1
            charSet.add(s[r])
            res = max(res, r - l + 1)
        return res

# Runtime1: 53ms, beats 68.95%
# Memory1: 16.54MB, beats 82.45%

# Runtime2: 52ms, beats 73.18%
# Memory2: 16.53MB, beats 82.45%

# Runtime3: 53ms, beats 68.95%
# Memory3: 16.56MB, beats 82.45%